src/main/java/uk/gov/ons/ctp/response/action/export/service/NotificationAndManifestFileCreator.java

```diff
@@ -51,34 +55,67 @@ public void uploadData(
       ExportJob exportJob,
       String[] responseRequiredList,
       int actionCount) {
+
     if (actionCount == 0) {
       return;
     }
 
-    final String now = FILENAME_DATE_FORMAT.format(clock.millis());
-    String filename = String.format("%s_%s.csv", filenamePrefix, now);
+    String now = FILENAME_DATE_FORMAT.format(clock.millis());
+    String csvFilename = String.format("%s_%s.csv", filenamePrefix, now);
+    String manifestFilename = csvFilename.replace(".csv", ".manifest");
 
-    if (exportFileRepository.existsByFilename(filename)) {
-      log.with("filename", filename)
-          .warn(
-              "Duplicate filename. The cron job is probably running too frequently. The "
-                  + "Action Exporter service is designed to only run every minute, maximum");
-      throw new RuntimeException();
-    }
+    checkForDuplicates(csvFilename, manifestFilename);
+    String directory = getDirectory(directorySuffix);
 
-    log.with("filename", filename).info("Uploading file");
+    writeFileToSftpAndRecordOnDB(
+        directory, csvFilename, exportJob, data, responseRequiredList, actionCount);
+
+    writeFileToSftpAndRecordOnDB(
+        directory,
+        manifestFilename,
+        exportJob,
+        manifestBuilder.createManifestData(csvFilename, data),
+        new String[0],
+        1);
+  }
+
+  private void writeFileToSftpAndRecordOnDB(
+      String directory,
+      String filename,
+      ExportJob exportJob,
+      ByteArrayOutputStream data,
+      String[] responseRequiredList,
+      int actionCount) {
 
+    log.with("filename", filename).info("Uploading file");
     ExportFile exportFile = new ExportFile();
     exportFile.setExportJobId(exportJob.getId());
     exportFile.setFilename(filename);
     exportFileRepository.saveAndFlush(exportFile);
+
+    sftpService.sendMessage(
+        directory, filename, responseRequiredList, Integer.toString(actionCount), data);
+    eventPublisher.publishEvent("Printed file " + filename);
+  }
+
+  private String getDirectory(String directorySuffix) {
     String directory = appConfig.getSftp().getDirectory();
     if (directorySuffix != null) {
       directory = directory.concat(directorySuffix).concat("/");
     }
-    sftpService.sendMessage(
-        directory, filename, responseRequiredList, Integer.toString(actionCount), data);
 
-    eventPublisher.publishEvent("Printed file " + filename);
+    return directory;
+  }
+
+  // Fudge for transactional management of rolling back SI's immediate sftp file commit
```

Comment:
Is this comment a little misleading? This piece of code is protection against multiple instances of the actionexporter trying to send files with the same name, if the cron wakes them up at more-or-less the same time, and the filenames end up being identical.

Reply:
fixed